minor: Add docs for blocks (#2217)
balena-io/docs: pages/learn/develop/blocks.md

---
title: Develop with blocks
excerpt: A drop-in chunk of functionality that can be added to your application to provide.
---

# Getting Started with blocks

> A software package is a collection of individual files or resources that are packed together to provide a certain functionality as part of a larger system. Blocks are just that. They are drop-in, independent chunks of functionality built to handle a specific task, allowing you to focus on solving the hard problems instead.

Author: I added independent on my own as I felt some blocks could be used independently of balena. Although most of them currently use some balenaAPI or feature. I will drop that. 

I talked about software packages first, because I feel it generates a sense of familiarity to the concept and thus making the understanding of a new concept "seem" easier than it actually is. It also felt like a nice hook into starting the docs for blocks. I can remove/modify it if you strongly feel that way.

> A software package is a collection of individual files or resources that are packed together to provide a certain functionality as part of a larger system. Blocks are just that. They are drop-in, independent chunks of functionality built to handle a specific task, allowing you to focus on solving the hard problems instead.

Reviewer: @vipulgupta2048 I think we can drop this paragraph and consolidate those thoughts into the below one, as per my next comment. It flows better.

> A block is a container image that can be added to a multi-container app in order to provide specific functionality. Blocks are intended to reduce the friction of application development by making it easier to re-use functionality, much like software packages. Similar to how software developers don't want to rewrite common functions, this is a DRY approach for IoT developers to not create containers of every function and re-use what's already out there for their application to work.

Reviewer: "....much like software packages provide functionality to a developer writing code".

Either link to an explanation of "DRY" or remove it. Currently we're tossing in a term people might not know.

I think the sentence beginning "Similar" needs a rewrite. Maybe:

"Similar to how software developers don't want to keep rewriting common functions, IOT developers don't want to keep creating containers for common tasks, such as handling Bluetooth connections or interacting with attached displays. Blocks provide solutions to common IOT application requirements, which developers can re-use to enable their Fleet."

Author: Love it, good point on removing DRY. Added complexity with not a lot of impact. Replace the line with your suggestion

For example, getting a browser to run well in a container is a tedious task. So, why not use the [browser block][browser].

<img src="/img/blocks/diagram-1.png" width="100%">

## Develop with a block

> Blocks being a container image can be added to a multi-container app. To use a block in your app, add the block's image as a new service in the `services` section of your app's `docker-compose.yml` file. This is similar to how we would [build multicontainer apps][multicontainer].

Reviewer: "As Blocks are a container image, they can be added to a multi-container Fleet."

Lots of use of app, when I think we want to keep to Fleet, to tie in with the mission and the words the user can see and use in the Dashboard.

Here's an example of using the [browser block][browser] in a sample multi-container app running 2 services. More blocks 

```
version: '2'

services:
  # Browser block service goes here 
  browser:
    image: balenablocks/browser
    privileged: true 
    network_mode: host

  # Existing services in the multi-container app
  balena-node-hello-world:
    build: ./balena-node-hello-world
    restart: always
    depends_on:
      - browser
    ports: 
      - "80"
  importantBananaDb:
    image: importantBanana/db:latest
    ports: 
      - "8080"
    environment:
      DB_ROOT_PASSWORD: 'potassium'
      RUN_MODE: 'super-secure'
``` 

> Similarly, each block will have be having their own set of [docker-compose fields][docker-compose] that it requires for configuration and usage. Do take care of syntax and conflicts in the docker-compose.yml file when composing and deploying your multi-container configuration. 

Reviewer: "Each Block may have their own set...."

More resources: 

1. Video tutorial on [IoT Happy Hour](https://youtu.be/Mllay6Z2-qQ?t=1440) about using  balenaBlocks and building an [environmental sensor app](https://github.com/balenalabs-incubator/ruuvitag) using sensor, connector, and dashboard block.
   
2. Check out our guide on [how to compose multi-container apps](https://www.balena.io/blog/two-projects-one-device-turn-your-raspberry-pi-into-a-multitool/) for more details. 
   
3. Blog on [Introduction to blocks](https://www.balena.io/blog/introducing-balenablocks-jumpstart-your-iot-app-development/).

## Creating your own block 

Since blocks are container images, the process to create blocks is very similar to creating a container image. To learn more about the what can be a block, you can read this [blog post](https://www.balena.io/blog/how-to-make-your-own-balenablocks-simple-drop-in-edge-app-functionality/). In order to create your own block:

1. The code for your block should be in GitHub project and has to [include a balena.yml file](https://www.balena.io/docs/learn/deploy/deploy-with-balena-button/#balenayml-configuration-file) and a logo. Here's a balena.yml file of a block for reference:

> ```yml

Reviewer: Is there a reason to show the capture block here, rather than stick to the browser block?

Author: No reason, I got this from the blog, I will swap this out.

---
title: Develop with blocks
excerpt: A drop-in chunk of functionality that can be added to your fleet.
---

# Develop with blocks

A block is a container image that can be added to a multicontainer fleet in order to provide specific functionality. Blocks are intended to reduce the friction of app development by making it easier to re-use functionality, much like software packages. 

Similar to how software developers don't want to keep rewriting common functions, IOT developers don't want to keep creating containers for common tasks, such as handling Bluetooth connections or interacting with attached displays. Blocks provide solutions to common IoT application requirements, which developers can re-use to enable their fleet.

> For example, getting a browser to run well in a container is a tedious task. Hence, why not use the [browser block][browser] to things up and running faster.

Reviewer: "to get things up and running faster."

<img src="/img/blocks/diagram-1.png" width="100%">

## Getting Started with blocks

> Blocks being a container image can be added to a multicontainer fleet. To use a block in your fleet, add the block's image as a new service in the `services` section of your app's `docker-compose.yml` file. This is similar to how we would build [multicontainer fleets][multicontainer].

Reviewer: "Since blocks are a container image, they can be added..."

Here's an example of using the [browser block][browser] in a sample multicontainer fleet already running 2 services.

```
version: '2'

services:
  # Browser block service added to the multicontainer fleet
  browser:
    image: balenablocks/browser
    privileged: true
    network_mode: host

  # Existing services in the multicontainer app
  balena-node-hello-world:
    build: ./balena-node-hello-world
    restart: always
    depends_on:
      - browser
    ports:
      - "80"
  importantBananaDb:
    image: importantBanana/db:latest
    ports:
      - "8080"
    environment:
      DB_ROOT_PASSWORD: 'potassium'
      RUN_MODE: 'super-secure'
```

> Each block will have be having their own set of [docker-compose fields][docker-compose] that's required for configuration and usage. Do take care of syntax and conflicts in the docker-compose.yml file when composing and deploying your multicontainer configuration.

Reviewer: "Each block will have their own...."
"which are required to configure the container they run within."

More resources:

1. Video tutorial on [IoT Happy Hour](https://youtu.be/Mllay6Z2-qQ?t=1440) about using  balenaBlocks and building an [environmental sensor app](https://github.com/balenalabs-incubator/ruuvitag) using sensor, connector, and dashboard block.
2. Check out our guide on [how to compose multicontainer fleets](https://www.balena.io/blog/two-projects-one-device-turn-your-raspberry-pi-into-a-multitool/) for more details.
3. Blog on [Introduction to blocks](https://www.balena.io/blog/introducing-balenablocks-jumpstart-your-iot-app-development/).

## Creating your own block

> Since blocks are container images, the process to create blocks is very similar to creating a container image. To learn more about the what can be a block, you can read this [blog post](https://www.balena.io/blog/how-to-make-your-own-balenablocks-simple-drop-in-edge-app-functionality/). 

Reviewer: "since blocks are container images, the process to create them is very similar to creating any other container. To learn more about what makes a good candidate for a block...."

> In order to create your own block. The code for your block should be in GitHub project. It has to [include a balena.yml file](https://www.balena.io/docs/learn/deploy/deploy-with-balena-button/#balenayml-configuration-file) and a logo. Here's a balena.yml file of the [browser block](https://github.com/balenablocks/browser/) for reference:

Reviewer: "in **a** GitHub project"

<img src="/img/blocks/create-block.png" width="70%">

Next, we have to [push a new release][balena-push] from your project to your newly created block on balenaCloud. If you are new to this, then follow the [Getting Started][getting-started] guide. After the release has been created, your block is ready. You can manage, modify and iterate further on your block from the balenaCloud dashboard.

## Release management for blocks

To deploy new releases of your block, you can use the [balena-push][balena-push] command using the [balenaCLI](https://github.com/balena-io/balena-cli). The `Releases` tab in the sidebar will show the list of past releases of your block as well as the release that is currently selected as default.

<img src="/img/blocks/release.png" width="80%">

To select a new default release for your block, click the `Default release` drop down menu in order to select a specific release from the list. When a particular release is selected, that release will be tagged as the default (or latest) release of your block.

<img src="/img/blocks/release-mgmt.png" width="100%">

## Making your block public

> Blocks when created on balenaCloud are by default set to private. Blocks when made public are neither joinable like fleets nor forkable like projects, but they’re a resource for everyone to use when creating their own fleets, projects, proof of concepts and prototypes.

Reviewer: "When a blocks is created on balenaCloud it will not be publicly visible on balenaHub. When a block is made public it is listed in the Blocks section of balenaHub, but is not joinable like fleets or forkable like projects. Instead the card for the block allows users to find the source code for the block, and learn how to deploy and use it in their own fleet."

> With blocks being private, you can iterate, test and build more functionality before making it available to the world. When you are ready to release, head to the `Settings` tab in the sidebar and click the `Make public` button to toggle the visibility of your block on balenaHub; you’re free to toggle the visibility on and off as necessary at any time.

Reviewer: "When blocks are private...."

---
title: Develop with blocks
excerpt: A drop-in chunk of functionality that can be added to your fleet.
---

# Develop with blocks

A block is a container image that can be added to a multicontainer fleet in order to provide specific functionality. Blocks are intended to reduce the friction of app development by making it easier to re-use functionality, much like software packages. 

Similar to how software developers don't want to keep rewriting common functions, IOT developers don't want to keep creating containers for common tasks, such as handling Bluetooth connections or interacting with attached displays. Blocks provide solutions to common IoT application requirements, which developers can re-use to enable their fleet.

For example, getting a browser to run well in a container is a tedious task. Hence, why not use the [browser block][browser] to get things up and running faster..

<img src="/img/blocks/diagram-1.png" width="100%">

## Getting Started with blocks

> Since blocks are a container images, they can be added to a multicontainer fleet. To use a block in your fleet, add the block's image as a new service in the `services` section of your app's `docker-compose.yml` file. This is similar to how we would build [multicontainer fleets][multicontainer].

Reviewer 2: Since blocks are container images*

Here's an example of using the [browser block][browser] in a sample multicontainer fleet already running 2 services.

```
version: '2'

services:
  # Browser block service added to the multicontainer fleet
  browser:
    image: balenablocks/browser
    privileged: true
    network_mode: host

>   # Existing services in the multicontainer app

Reviewer 2: Just checking, should this be app or fleet? Because above it says fleet

  balena-node-hello-world:
    build: ./balena-node-hello-world
    restart: always
    depends_on:
      - browser
    ports:
      - "80"
  importantBananaDb:
    image: importantBanana/db:latest
    ports:
      - "8080"
    environment:
      DB_ROOT_PASSWORD: 'potassium'
      RUN_MODE: 'super-secure'
```

Each block will have their own set of [docker-compose fields][docker-compose] which are required to configure the container they run within. Do take care of syntax and conflicts in the docker-compose.yml file when composing and deploying your multicontainer configuration.

More resources:

1. Video tutorial on [IoT Happy Hour](https://youtu.be/Mllay6Z2-qQ?t=1440) about using  balenaBlocks and building an [environmental sensor app](https://github.com/balenalabs-incubator/ruuvitag) using sensor, connector, and dashboard block.
2. Check out our guide on [how to compose multicontainer fleets](https://www.balena.io/blog/two-projects-one-device-turn-your-raspberry-pi-into-a-multitool/) for more details.

> 3. Blog on [Introduction to blocks](https://www.balena.io/blog/introducing-balenablocks-jumpstart-your-iot-app-development/).

Reviewer 2: Shouldn't these 2 come before the above step? Sounds like knowing how to compose multicontainers and being introduced to blocks should be pre-requisites of a tutorial on using blocks

## Creating your own block

Since blocks are container images, the process to create them is very similar to creating any other container. To learn more about what makes a good candidate for a block, refer to this [blog post](https://www.balena.io/blog/how-to-make-your-own-balenablocks-simple-drop-in-edge-app-functionality/). 

> In order to create your own block. The code for your block should be in a GitHub project. It has to [include a balena.yml file](https://www.balena.io/docs/learn/deploy/deploy-with-balena-button/#balenayml-configuration-file) and a logo. Here's a balena.yml file of the [browser block](https://github.com/balenablocks/browser/) for reference:

Reviewer 2: In order to create your own block, the code for your block should be in a GitHub project.*

Reviewer 2: Also I'm not certain a logo is _required_, we have a default logo that appears if the user-provided on does not work. Double-check with @chrisys or someone whether a logo is really required when creating a block

Author: Removed the logo requirement

  connected display
version: 2.3.2
type: sw.block
assets:
  repository:
    type: blob.asset
    data:
      url: 'https://github.com/balenablocks/browser'
  logo:
    type: blob.asset
    data:
      url: 'https://raw.githubusercontent.com/balenablocks/browser/master/logo.png'
data:
  defaultDeviceType: raspberrypi3
  supportedDeviceTypes:
    - raspberrypi4-64
    - fincm3
    - raspberrypi3
    - raspberrypi3-64
    - raspberrypi400-64
    - intel-nuc
    - genericx86-64-ext
```

Next, navigate to the `Blocks` tab in the sidebar and click the `Create block` button in the balenaCloud dashboard.

<img src="/img/blocks/image.png" width="100%">

Fill the details for your block in the modal, and click `Create new block` button to create your block.

<img src="/img/blocks/create-block.png" width="70%">

Next, we have to [push a new release][balena-push] from your project to your newly created block on balenaCloud. If you are new to this, then follow the [Getting Started][getting-started] guide. After the release has been created, your block is ready. You can manage, modify and iterate further on your block from the balenaCloud dashboard.

## Release management for blocks

> To deploy new releases of your block, you can use the [balena-push][balena-push] command using the [balenaCLI](https://github.com/balena-io/balena-cli). The `Releases` tab in the sidebar will show the list of past releases of your block as well as the release that is currently selected as default.

Reviewer 2: Why not just:
```suggestion
The `Releases` tab in the sidebar will show the list of all releases of your block.
```

<img src="/img/blocks/release.png" width="80%">

> By default, the block set to `track latest`, which means that new releases are immediately tagged as the default release for the block. Expanding this _Default release_ dropdown menu displays all releases for the block, and you can select a specific release to pin the block to. When a particular release is selected, that release will be tagged as the default (or latest) release of your block. In future, releases will not be deployed until the _Default release_ is updated to a newer release or set back to `track latest`.

Reviewer 2: By default, the block is set to*

Reviewer 2: Expanding this _Default release_ dropdown menu displays all successful releases for the block*

Reviewer 2: > default (or latest)

I don't think the `(or latest)` part is true or that we should use that phrasing 😓

Reviewer 2: In the future, releases will*

<img src="/img/blocks/release-mgmt.png" width="100%">

## Making your block public

When a block is created on balenaCloud it will not be publicly visible on balenaHub. When a block is made public it is listed in the [Blocks section][blockhub] of balenaHub, but is not joinable like fleets or forkable like projects. Instead the card for the block allows users to find the source code for the block, and learn how to deploy and use it in their own fleet.

> When blocks are private, you can iterate, test and build more functionality before making it available to the world. When you are ready to release, head to the `Settings` tab in the sidebar and click the `Make public` button to toggle the visibility of your block on balenaHub; you’re free to toggle the visibility on and off as necessary at any time.

Reviewer 2: As Phil mentioned in FD, there's no `Make public` button. I know you're aware of this but just leaving this comment so we don't forget to update this.

Also we should specify here that a block cannot be made publicly visible unless there is a repo URL provided (even though this won't be a _requirement_ anymore in the future, rather just an option. When that happens we'll need to update the docs again)
